issue #39 - early work on updating the log display

bfitzpat opened the pull request:
Signed-off-by: bfitzpat@redhat.com <bfitzpat@redhat.com>

bfitzpat commented:
![image](https://user-images.githubusercontent.com/530878/74289110-67c1c880-4ceb-11ea-9252-81a6f63f88b1.png)

bfitzpat commented:
And yes, there will be some errors here. Just made sure to get it stashed so we could chat about it.

bfitzpat commented:
So I now have the webview log working the same as it does in the Kubernetes extension. You can open it and move it around. It suffers from the same problem as the original, not scrolling to the bottom appropriately to follow the log -- so we may need to figure something out there.

Only weirdness is the failing kubectlwatcher tests @apupier  -- I'm not quite sure what those are doing. Any insights into why those would be failing?

bfitzpat commented:
Another thing we now have to figure out is how best to error out when the pod doesn't exist for a log to follow. Right now we show a note in the Apache Camel K output channel, but that may not be enough.

![image](https://user-images.githubusercontent.com/530878/74385359-9d7cb500-4db0-11ea-9dab-20c6420765e0.png)

bfitzpat commented:
And then, when we get it working right, we need to look at how to format it so that errors and warnings are presented nicely in colors other than that drab brown.

bfitzpat commented:
But we can now launch the Camel K Operator log from the Kubernetes extension views and the Camel K integration logs from the Camel K integrations view and get something like this...

![image](https://user-images.githubusercontent.com/530878/74386881-54c6fb00-4db4-11ea-9d59-518d62dcf7b0.png)

bfitzpat commented:
And need to close the log when we remove the integration. That's another thing left to do.

apupier commented:
> It suffers from the same problem as the original, not scrolling to the bottom appropriately to follow the log -- so we may need to figure something out there.

this issue https://github.com/Azure/vscode-kubernetes-tools/issues/593 , right?

bfitzpat commented:
Riccardo gave me the scroll-to-bottom solution, so though that's not the BEST solution it at least updates the log so you can see the latest updates as it scrolls past.

apupier commented:
please try to provide the fix upstream first to avoid duplicated code that diverge even before the first merge

bfitzpat commented:
> I guess the provided solution is not covering the issue mentioned int his comment [Azure/vscode-kubernetes-tools#593 (comment)](https://github.com/Azure/vscode-kubernetes-tools/issues/593#issuecomment-516672587) about

You are correct. It does not solve that problem. But it does scroll to the bottom and continuously show the log. I at this point do not know how to turn that off, but I suppose maybe we can add a checkbox to turn it off?

bfitzpat commented:
> please try to provide the fix upstream first to avoid duplicated code that diverge even before the first merge

I want to comment on this specific line. Though we can contribute things upstream (and I plan to since this was an easy option), since we can't reuse the code directly in VS Code we won't be able to take advantage of that change. And if we change things to add colors for warning/error lines and alter the view further, it's going to diverge very quickly. 

But since the codebases can't be shared due to limitations in extension architecture, our code is going to duplicate upstream effort and diverge regardless.

bfitzpat commented:
I have added an auto-scroll checkbox to the header on the view that turns that feature on/off and defaults to on. It works pretty well. I think I'm going to remove the filter controls though, since they don't currently do anything. If they get those to work upstream, we can collect that functionality as well -- but until then they just add clutter. 

![image](https://user-images.githubusercontent.com/530878/74462445-7b854000-4e4d-11ea-954c-b1df833de9c4.png)

bfitzpat commented:
Removed the filter fields from our log viewer until they get them working upsream. But here's a screen shot of a Kubernetes-provided log from the camel k operator and one of ours from Camel K. 

![image](https://user-images.githubusercontent.com/530878/74463231-c2c00080-4e4e-11ea-8a58-768cc3524341.png)

bfitzpat commented:
Proposed our solution upstream - https://github.com/Azure/vscode-kubernetes-tools/issues/593#issuecomment-585893634 and https://github.com/Azure/vscode-kubernetes-tools/pull/704

bfitzpat commented:
I have a weird issue I'm trying to diagnose... If you start a log and then move it to another column/location, we lose the content from the beginning and then it starts again. I suspect it's due to the script in the logsWebview.ts - but I honestly can't figure it out.

apupier commented:
> > please try to provide the fix upstream first to avoid duplicated code that diverge even before the first merge
> 
> I want to comment on this specific line. Though we can contribute things upstream (and I plan to since this was an easy option), since we can't reuse the code directly in VS Code we won't be able to take advantage of that change. And if we change things to add colors for warning/error lines and alter the view further, it's going to diverge very quickly.
> 
> But since the codebases can't be shared due to limitations in extension architecture, our code is going to duplicate upstream effort and diverge regardless.

if we can keep code duplicated but exactly the same it will be easier to apply fix/improvements done upstream to our codebase. Otherwise it will require potentially conflicting change to backport from upstream to our branch.

for colors for warnings and errors, I think that this is a feature that will be useful upstream too.

Which very specific modifications to Camel K (meaning that won't benefit to upstream too) do you see?

bfitzpat commented:
> if we can keep code duplicated but exactly the same it will be easier to apply fix/improvements done upstream to our codebase. Otherwise it will require potentially conflicting change to backport from upstream to our branch.

If the code upstream was more mature, I might agree with this. The filter fields in the view have been there in an ornamental state since July 2019. _Duplicating non-working code from upstream and keeping it exactly the same even if it doesn't work doesn't help us._ As a result, we are inevitably going to diverge.

bfitzpat commented:
Ok, fixed the issue with the log being cut off. Had to do with the two fields for filtering I removed. Enabled those and now it seems to be fine. I think this is a good initial commit for this functionality, which removes the output channel functionality and shifts these logs to individual webview-powered windows that can be moved around the vscode interface. 

We will have to investigate the idea of "pretty print" for the log output, which can likely be handled as we push each log line (after parsing by newlines) to the webview. We can do that separately.

bfitzpat commented:
> We will have to investigate the idea of "pretty print" for the log output, which can likely be handled as we push each log line (after parsing by newlines) to the webview. We can do that separately.

If we're truly pushing for a duplication of the code from the Kubernetes extension, perhaps log formatting is less important at this point

apupier commented on src/extension.ts:
```diff
@@ -417,46 +404,41 @@ export function shareMessageInMainOutputChannel(msg: string) {
+		const panel = LogsPanel.createOrShow('Loading...', cresource);
+
+		let cmd = `logs -f ${podName}`;
+        await kubectlExe.invokeAsync(cmd, undefined, (proc: ChildProcess) => {
```
indentation issue

apupier commented on src/extension.ts:
```diff
@@ -417,46 +404,41 @@ export function shareMessageInMainOutputChannel(msg: string) {
+		let cmd = `logs -f ${podName}`;
+        await kubectlExe.invokeAsync(cmd, undefined, (proc: ChildProcess) => {
+			if (proc && proc.stdout) {
+            	proc.stdout.on('data', (data: string) => {
```
indentation issue

apupier commented on src/kubectl.ts:
```diff
@@ -44,6 +47,62 @@ class KubectlImpl implements Kubectl {
+		let sr: ShellResult | undefined;
+		if (stdin) {
+			sr = await exec(cmd, stdin);
+        } else if (callback) {
```
indentation issue

apupier commented on src/kubectl.ts:
```diff
@@ -44,6 +47,62 @@ class KubectlImpl implements Kubectl {
+    try {
+        return await execCore(cmd, null, callback);
+    } catch (ex) {
+        console.log(ex);
```
Why not sending to the outputchannel?

bfitzpat replied:
Fixed

apupier commented on src/kubectl.ts:
```diff
@@ -44,6 +47,62 @@ class KubectlImpl implements Kubectl {
+    try {
+        return await execCore(cmd, null, null, stdin);
+    } catch (ex) {
+        console.log(ex);
```
Why not sending to the outputchannel?

bfitzpat replied:
Fixed

apupier commented on src/kubectl.ts:
```diff
@@ -44,6 +47,62 @@ class KubectlImpl implements Kubectl {
+}
+
+export interface ShellResult {
+    readonly code: number;
```
space instead fo tabs used

apupier commented on src/kubectl.ts:
```diff
@@ -44,6 +47,62 @@ class KubectlImpl implements Kubectl {
+}
+
+async function exec(cmd: string, stdin?: string): Promise<ShellResult | undefined> {
+    try {
```
space instead fo tabs used

apupier commented on src/kubectl.ts:
```diff
@@ -44,6 +47,62 @@ class KubectlImpl implements Kubectl {
+}
+
+function execCore(cmd: string, opts: any, callback?: ((proc: ChildProcess) => void) | null, stdin?: string): Promise<ShellResult> {
+    return new Promise<ShellResult>((resolve) => {
```
space instead fo tabs used

apupier commented on src/logsWebview.ts:
```diff
@@ -0,0 +1,172 @@
+import * as vscode from 'vscode';
```
missing file header

apupier commented on src/webpanel.ts:
```diff
@@ -0,0 +1,78 @@
+import * as vscode from 'vscode';
```
missing header

bfitzpat replied:
Added MIT license header and mention that we got this from the kubernetes tooling project

apupier commented on testFixture/.vscode/settings.json:
```diff
@@ -0,0 +1,10 @@
+{
```
i think this shouldn't be commited

apupier commented on src/logsWebview.ts:
```diff
@@ -0,0 +1,172 @@
+            </script>
+            </body>
+        </html>`;
+    }
```
wrong indentation level

apupier commented on src/logsWebview.ts:
```diff
@@ -0,0 +1,172 @@
+            </body>
+        </html>`;
+    }
+}
```
wrong indentation level

apupier replied:
there is still a wrong indentation level. The indentation is correct upstream https://github.com/Azure/vscode-kubernetes-tools/blob/a7d45e92e39036256c610cec9cb0245c96453fb2/src/components/logs/logsWebview.ts#L161

apupier replied:
this is fine now (I should have put my comment on the line above so that it would have marked my comments as outdated automatically)

apupier commented on src/logsWebview.ts:
```diff
@@ -0,0 +1,172 @@
+
+    protected update() {
+        if (this.panel.visible) {
+        this.panel.title = `Logs - ${this.resource}`;
```
wrong indetantion level

apupier commented:
> > We will have to investigate the idea of "pretty print" for the log output, which can likely be handled as we push each log line (after parsing by newlines) to the webview. We can do that separately.
> 
> If we're truly pushing for a duplication of the code from the Kubernetes extension, perhaps log formatting is less important at this point

i think we should provide the feature upstream first. this is not a specific feature for Camel K needs

lhein commented:
> most of the commit messages are missing the issue number

I think we can do a squash commit with a corrected issue number. Its just 8 files and not worth the hassle to rewrite the commit history.

apupier commented on src/extension.ts:
```diff
@@ -100,10 +98,7 @@ export async function activate(context: vscode.ExtensionContext): Promise<void>
-				.catch( (err) => {
-					console.log(err);
-				});
+				// TODO: we need to look into closing the log view when the integration is stopped
```
have you created a specific issue yet?

lhein replied:
https://github.com/camel-tooling/vscode-camelk/issues/310

lhein replied:
I also created https://github.com/camel-tooling/vscode-camelk/issues/311 to track the issue of the stale log views when updating an integration changes the pod.

apupier commented:
> > most of the commit messages are missing the issue number
> 
> I think we can do a squash commit with a corrected issue number. Its just 8 files and not worth the hassle to rewrite the commit history.

this is an hassle of 20-30 seconds for a simple squash... which allows to review the commit message and history. But already discussed directly with Brian which doesn't see the value either.

lhein commented:
@apupier I think we can agree on doing a squash into a single commit and force push that one with a correct issue number and title?

apupier commented:
> @apupier I think we can agree on doing a squash into a single commit and force push that one with a correct issue number and title?

this is the minimal solution that I proposed.

bfitzpat commented on README.md:
```diff
@@ -218,13 +218,13 @@ If you want to explicitly view the running log for a published integration in a
+![Apache Camel K Integrations view - Integration log](images/camelk-integrations-view-integrations-log-revised.jpg)
 
-## Stopping an Apache Camel K integration
+Note: By default, `Auto-scroll` is checked and new entries in the log will automatically be scrolled to at the bottom of the window. If turned off, the scrolling stops and can be done manually.
```
I disagree. The sentence makes sense as is. "new entries in the log will automatically be scrolled to at the bottom of the window". Leaving it.

lhein replied:
You are the native speaker. Fine for me.

apupier replied:
fine for me too. you're the native speaker.
I reported because it feels strange to have two prepositions and Google doc autocorrection is suggesting to remove the 'at'
